source/common/upstream/cluster_manager_impl.cc

```diff
@@ -377,21 +394,28 @@ ClusterManagerImpl::ClusterManagerImpl(
       primary_clusters_.insert(cluster.name());
     }
   }
+
+  bool has_ads_cluster = false;
   // Load all the primary clusters.
   for (const auto& cluster : bootstrap.static_resources().clusters()) {
     if (is_primary_cluster(cluster)) {
-      loadCluster(cluster, MessageUtil::hash(cluster), "", false, active_clusters_);
+      const bool required_for_ads = isAdsCluster(bootstrap, cluster.name());
+      has_ads_cluster |= required_for_ads;
+      loadCluster(cluster, MessageUtil::hash(cluster), "", /*added_via_api=*/false,
```

Comment:
One alternative here is to give each cluster a post-load/initialization callback which we then pass the ADS start continuation to. I don't think it's strictly necessary but throwing it out there in case there are other use cases I'm missing.

Reply:
This is an interesting idea and thought about it myself, but then wasn't really sure if it was worth adding the extra logic in as I didn't see any other use cases.  I think I'll add a note that if we end up finding we need to do more of this "post cluster init" stuff, it would be good to build the ClusterManagerImpl interfaces to support it.  WDYT?

Reply:
Yes, that's fine, this was just a suggestion, fine to take a note on this.